Retrying Requests when an IOError/OSError exception is raised when decompressing Response (#4096)
scrapy/scrapy: docs/topics/downloader-middleware.rst

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. setting:: COMPRESSION_ENABLED

> .. setting:: RETRY_HTTP_COMPRESSION_ERRORS

Reviewer: This line should be right above the `RETRY_HTTP_COMPRESSION_ERRORS` section instead :slightly_smiling_face:

Author: My bad 🤦‍♂

Whether the Compression middleware will be enabled.

RETRY_HTTP_COMPRESSION_ERRORS
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Default: ``False``

> Whether we should retry requests when an exception is raised while trying

Reviewer: :lipstick: Style suggestions:
* I think in English “whether” only makes sense if there is an “or” somewhere. You can alternatively use “if” instead.
* I would avoid “we” here, and instead switch to passive.
* Since you use “its” in the next line, using singular “request” makes more sense; you could also change “its” to “theirs” instead, but I find that using singular in this context is more common in English.

```suggestion
Whether or not a request should be retried when an exception is raised while trying
```

Author: I'm used to native speakers omitting the _or not_ part of the _whether_ expression as it's most of the time redundant.

Here you can find some references:
- https://dictionary.cambridge.org/dictionary/english/whether
- https://english.stackexchange.com/questions/3382/whether-or-not-vs-whether

It's also being used this way by the `COMPRESSION_ENABLED` setting definition above.

In any case, I do find your suggested change better and I'm willing to commit it. Thank you.

Reviewer: Thanks for the links! Good to know :smiley:

to decompress its response body.

> If an exception is raised while trying to decompress the response body,

Reviewer: :lipstick: 
```suggestion
If an exception is raised while trying to decompress a response body,
```

> we store it in the request meta under the ``_http_compression_exc`` key and

Reviewer: ```suggestion
that exception is stored in the request :attr:`~scrapy.http.Request.meta` under the ``_http_compression_exc`` key and
```

return the original ``scrapy.Response`` object. If the ``RetryMiddleware`` is

> enabled, it will check for this key and try to reprocess the ``scrapy.Request``

Reviewer: ```suggestion
enabled, it will check for this key and try to reprocess the :class:`~scrapy.http.Request`
```

if the captured exception is specified in the

> ``RetryMiddleware.EXCEPTIONS_TO_RETRY`` tuple.

Reviewer: I think we should either remove this “if …” statement from the documentation or expose `EXCEPTIONS_TO_RETRY` through a setting.

Author: I do agree with you. Although I prefer to expose this class attribute through a setting, this is out of the scope of this ticket. So I'm just removing the _if_.
